readme.md
You are in the iteration-3 branch. Make sure to checkout to the correct iteration branch before making changes!

changelog.txt

#---------------#
#	Changelog	#
#---------------#



#-------------------------------#
#	Update: October 15, 2017	#
#-------------------------------#

-Added this .txt file.

-in app.py:
	-added 'db.session.refresh(account)'	//this should fix an error where a new account did not appear in the database until the server was restarted
	-changed "gpa" to a string temporarily; will change back to float value after string->float conversion code is added

-in my-account.html:
	-changed the text a bit; no changes to the code itself were made.
	
-in talink.js:
	-added the functions:
		-makeGetRequest
		-makePostRequest
		-attachCreateAccountHandler
		-createStudentAccount
		-createInstructorAccount
		-attachLoginHandler

testing/testLib.py
"""
A library for functional testing of the backend API
"""

import unittest
import traceback
import http.client
import sys
import os
import json


class RestTestCase(unittest.TestCase):
    """
    Superclass for functional tests. Defines the boilerplate for the tests
    """

    # Lookup the name of the server to test
    serverToTest = "localhost:5000"
    accountSpace = 'testingSpaceAuto'
    VERBOSE = 0


    def makeRequest(self, url, method="GET", data={ }):
        """
        Make a request to the server.
        @param url is the relative url (no hostname)
        @param method is either "GET" or "POST"
        @param data is an optional dictionary of data to be send using JSON
        @result is a dictionary of key-value pairs
        """

        printHeaders = (self.VERBOSE == 1)
        headers = { }
        body = ""
        if data is not None:
            headers = { "content-type": "application/json", "Accept": "application/json" }
            body = json.dumps(data)

        try:
            self.conn.request(method, url, body, headers)
        except Exception as e:
            if str(e).find("Connection refused") >= 0:
                print ("Cannot connect to the server ",RestTestCase.serverToTest,". You should start the server first, or pass the proper TEST_SERVER environment variable")
                sys.exit(1)
            raise

        self.conn.sock.settimeout(100.0) # Give time to the remote server to start and respond
        resp = self.conn.getresponse()
        data_string = "<unknown"
        try:
            if printHeaders:
                print("\n****")
                print("  Request: ",method," url=",url," data=",str(data))
                print("  Request headers: ",str(headers))
                print("")
                print("  Response status: ",str(resp.status))
                print("  Response headers: ")
                for h, hv in resp.getheaders():
                    print("    ",h,"  =  ",hv)

            self.assertEqual(200, resp.status)
            data_string = resp.read().decode()
            if printHeaders:
                print("  Response data: ",data_string)
            # The response must be a JSON request
            # Note: Python (at least) nicely tacks UTF8 information on this,
            #   we need to tease apart the two pieces.
            self.assertTrue(resp.getheader('content-type') is not None, "content-type header must be present in the response")
            self.assertTrue(resp.getheader('content-type').find('application/json') == 0, "content-type header must be application/json")


            data = json.loads(data_string)
            return data

        except:
            # In case of errors dump the whole response,to simplify debugging
            print("Got exception when processing response to url="+url+" method="+method+" data="+str(data))
            print("  Response status = "+str(resp.status))
            print("  Response headers: ")
            for h, hv in resp.getheaders():
                print("    "+h+"  =  "+hv)
            print("  Data string: ",data_string)
            print("  Exception: ",traceback.format_exc ())
            if not printHeaders:
                print("  If you want to see the request headers, use VERBOSE=1")


    def setUp(self):
        self.conn = http.client.HTTPConnection(RestTestCase.serverToTest, timeout=1)

    def tearDown(self):
        self.conn.close ()

		
		
		

## NOW SOME METHODS THAT ARE USEFUL FOR TESTING

class AccountTestCase(RestTestCase):
    """
    A base class for testing accounts
    """

    def setUp(self):
        # Run first the setUp from the superclass
        RestTestCase.setUp(self)
        # We empty the smile space, before we start the test
        self.emptyAccountSpace()

	
    def getAccounts(self):
        """
        Helper function to gets all accounts
        """
		
        url = '/api/account/getAll?space=' + self.accountSpace
        print(url)

        respData = self.makeRequest(url, method='GET')
        return respData


    def emptyAccountSpace(self):
        """
        Helper function to delete an account space
        """
        url = '/api/account?space='+self.accountSpace
        self.makeRequest(url, method='DELETE')

		
    def assertSuccessResponse(self,
                              respData,
                              msg=None):
        """
        Check that the response is not an error response
        """
        self.assertTrue(respData != None, 'Response is None')
        self.assertEqual(1, respData['status'], msg)

static/scripts/talink.js


var TALink = (function(){
	
	//----------------------Place private variables here-----------------------
	
	var apiUrl = 'http://localhost:5000';	//	where the backend is being hosted; right now, it's on your local machine
	
	var accountSpace = 'testingSpace';	//	this is the space in which the accounts we create live; 
										//	a tag to separate them from other sets of accounts.
	
	
	//-------------------------------------------------------------------------
	
	
	//----------------------------Private methods:-----------------------------
	
	/**
    * HTTP GET request 
    * @param  {string}   url       URL path, e.g. "/api/account"
    * @param  {function} onSuccess   callback method to execute upon request success (200 status)
    * @param  {function} onFailure   callback method to execute upon request failure (non-200 status)
    * @return {None}
    */
   var makeGetRequest = function(url, onSuccess, onFailure) {
       $.ajax({
           type: 'GET',
           url: apiUrl + url,
           dataType: "json",
           success: onSuccess,
           error: onFailure
       });
   };

     /**
     * HTTP POST request
     * @param  {string}   url       URL path, e.g. "/api/account"
     * @param  {Object}   data      JSON data to send in request body
     * @param  {function} onSuccess   callback method to execute upon request success (200 status)
     * @param  {function} onFailure   callback method to execute upon request failure (non-200 status)
     * @return {None}
     */
    var makePostRequest = function(url, data, onSuccess, onFailure) {
        $.ajax({
            type: 'POST',
            url: apiUrl + url,
            data: JSON.stringify(data),
            contentType: "application/json",
            dataType: "json",
            success: onSuccess,
            error: onFailure
        });
    };
	
	var homeLoadUserData = function(){
		var pathnameSplit = window.location.pathname.split("/");
		var pathname = pathnameSplit[pathnameSplit.length - 1];
		
		if (pathname == "home.html"){
			console.log("confirmed on home.html!");
			
			console.log(localStorage.getItem("user_type"));
			console.log(localStorage.getItem("user_type") == "Instructor");
			
			displayUserTypeClasses();
			
			$("#user-name").text(localStorage.getItem("first_name") + ' ' + localStorage.getItem("last_name"));
			$("#student-major").text(localStorage.getItem("major"));
			$("#student-graduation-date").text(localStorage.getItem("expected_grad"));
			$("#student-TA-history").text(localStorage.getItem("ta_before"));
		}
	};
	
	var accountLoadUserData = function(){
		var pathnameSplit = window.location.pathname.split("/");
		var pathname = pathnameSplit[pathnameSplit.length - 1];
		
		if (pathname == "account.html"){
			displayUserTypeClasses();
			
			$("#first-name").val(localStorage.getItem("first_name"));
			$("#last-name").val(localStorage.getItem("last_name"));
			$("#wsu-email").val(localStorage.getItem("username"));
			$("#wsu-id").val(localStorage.getItem("wsu_id"));
			$("#phone").val(localStorage.getItem("phone_number"));
			$("#personal-email").val(localStorage.getItem("secondary_email"));
			$("#major").val(localStorage.getItem("major"));
			$("#gpa").val(localStorage.getItem("gpa"));
			
			//Split the "expected_grad" string (example of returned string: "Fall 2017")
			var expectedGradSplitList = localStorage.getItem("expected_grad").split(" ");
			var expectedGradSemester = expectedGradSplitList[0];
			var expectedGradYear = expectedGradSplitList[1];
			
			console.log(expectedGradSemester);
			console.log(expectedGradYear);
			console.log(localStorage.getItem("ta_before"));
			
			//Check the correct radio button for expected grad semester
			if (expectedGradSemester == "Fall"){
				$("input:radio[name=graduation-semester]").filter('[value=Fall]').prop('checked', true);
			}
			
			else if (expectedGradSemester == "Spring"){
				$("input:radio[name=graduation-semester]").filter('[value=Spring]').prop('checked', true);
			}
			
			//Fill in the expected graduation year field
			$("#graduation-year").val(expectedGradYear);
			
			//Check the correct radio button for TA history
			if (localStorage.getItem("ta_before") == "true"){
				$("input:radio[name=ta-prior]").filter('[value=yes]').prop('checked', true);
			}
			
			else{
				$("input:radio[name=ta-prior]").filter('[value=no]').prop('checked', true);
			}
		}
		
	};
	
	var displayUserTypeClasses = function(){
		if(localStorage.getItem("user_type") == "Instructor"){
			$('.instructor').each(function () {
				this.style.setProperty('display', 'initial', '');
			});
			
			$('.student').each(function () {
				this.style.setProperty('display', 'none', '');
			});
			
			//This doesn't apply to home.html, but we'll just cover everything here.
			$('.form-student').each(function () {
				this.style.setProperty('display', 'none', '');
			});
		}
	};
	
	//	Attach an event listener to the account type radio that handles making
	//	student-specific attributes visible or not.
	//	Also handles making sure that fields not shown are not required for
	//	form submission.
	var attachStudentFieldDisplayListener = function(e){
		$(document).on("click", "input[name=account-type]", function(e){
			if ($("input[name=account-type]:checked").val() == "student"){
				$(".form-student").css("display", "block");
				$(".form-student input").prop("required", true);
			}
			
			else if($("input[name=account-type]:checked").val() == "instructor"){
				$(".form-student").css("display", "none");
				$(".form-student input").prop("required", false);
			}
			
			else{
				alert("Error! Not a student or instructor?");
			}
		});
	};
	
	
	
	//	Without this function, radios will not reset to their default value after
	//	page refresh. This causes an issue where instructor can be selected with
	//	student attributes visible, since those are visible by default.
	var resetRadios = function(){
		$(".account-creation-form").each(function(){
			this.reset();
		});
	};
	
	
	
	//	Check to see if the password fields are the same. Called by the event 
	//	listeners placed on #password and #confirm-password by 
	//	attachConfirmPasswordHandler() below.
	var confirmPassword = function(){
		if($("#password").val() == $("#confirm-password").val()){
			$("#confirm-password")[0].setCustomValidity("");
		}
		
		else{
			$("#confirm-password")[0].setCustomValidity("Passwords do not match.");
		}
	};
	
	
	
	//	Attach event listeners to #password and #confirm-password that call the 
	//	confirmPassword() function when a key is released in those fields.
	var attachConfirmPasswordListener = function(){
		$(document).on("keyup", "#password", function(e){
			confirmPassword();
		});
		
		$(document).on("keyup", "#confirm-password", function(e){
			confirmPassword();
		});
	};
	
	var createStudentAccount = function(){
		
		//cFirstName = $(".account-creation-form").find('first-name').val();
		//cLastName = $(".account-creation-form").find('last-name').val();
		//cUsername = $(".account-creation-form").find('wsu-email').val();
		
		var accountInfo = {};	//prepare the account object to send to the server
		//And now we fill the object with information from the forms
		accountInfo.user_type = "Student";
		accountInfo.space = accountSpace;
		accountInfo.first_name = $('#first-name').val();
		accountInfo.last_name = $('#last-name').val();
		accountInfo.wsu_id = $('#wsu-id').val();
		accountInfo.wsu_email = $('#wsu-email').val();
		accountInfo.password = $('#password').val();
		accountInfo.phone_number = $('#phone').val();
		accountInfo.secondary_email = $('#personal-email').val();
		accountInfo.major = $('#major').val();
		accountInfo.gpa = $('#gpa').val();
		accountInfo.course_preferences = [];
		
		if ($("input[name=ta-prior]:checked").val() == "yes"){
			accountInfo.ta_before = true;
		}
		else{
			accountInfo.ta_before = false;
		}
		
		if ($("input[name=ta-prior]:checked").val() == "spring"){
			accountInfo.expected_grad = "Spring " + $('#graduation-year').val();
		}
		else{
			accountInfo.expected_grad = "Fall " + $('#graduation-year').val();
		}
		
        var onSuccess = function(data) {  
			window.location.href = "index.html";	//if we successfully created an account, go back to the login page
        };
        var onFailure = function() { 
            console.error('create account failed'); 
        };
		
		//make a post request, supplying the accountInfo object we just filled out
		makePostRequest('/api/account/student', accountInfo, onSuccess, onFailure);
		
		
	};
	
	var createInstructorAccount = function(){
	
		//And now we fill the object with information from the forms
		var accountInfo = {};	//prepare the account object to send to the server
		accountInfo.user_type = "Instructor";
		accountInfo.space = accountSpace;
		accountInfo.first_name = $('#first-name').val();
		accountInfo.last_name = $('#last-name').val();
		accountInfo.wsu_id = $('#wsu-id').val();
		accountInfo.wsu_email = $('#wsu-email').val();
		accountInfo.password = $('#password').val();
		accountInfo.phone_number = $('#phone').val();
		accountInfo.secondary_email = $('#personal-email').val();
		accountInfo.courses_taught = [];
	
	
	    var onSuccess = function(data) {
			window.location.href = "index.html";	//if we successfully created an account, go back to the login page
        };
        var onFailure = function() { 
            console.error('create account failed'); 
        };
		
		//make a post request, supplying the accountInfo object we just filled out
		makePostRequest('/api/account/instructor', accountInfo, onSuccess, onFailure);
	
	};
	
	var attachLoginHandler = function(e){
		
		$(".login-form").on('click', ".login-button", function(e) {
			e.preventDefault ();	// Tell the browser to skip its default click action
		
			var lUsername = $(".login-form").find('.username-input').val();
			var lPassword = $(".login-form").find('.password-input').val();
		
			var onSuccess = function(data) {
				window.location.href = "home.html";	//if we successfully logged into an account, go to the account page

				window.localStorage.setItem("user_type", data["person"]["user_type"]);
				
				window.localStorage.setItem("username", data["person"]["wsu_email"]);
				window.localStorage.setItem("major", data["person"]["major"]);
				window.localStorage.setItem("expected_grad", data["person"]["expected_grad"]);
				window.localStorage.setItem("ta_before", data["person"]["ta_before"]);
				window.localStorage.setItem("first_name", data["person"]["first_name"]);
				window.localStorage.setItem("last_name", data["person"]["last_name"]);
				window.localStorage.setItem("wsu_id", data["person"]["wsu_id"]);
				window.localStorage.setItem("phone_number", data["person"]["phone_number"]);
				window.localStorage.setItem("secondary_email", data["person"]["secondary_email"]);
				window.localStorage.setItem("gpa", data["person"]["gpa"]);
			};
			var onFailure = function() { 
				console.error('login failed'); 
			};
			//make a get request, supplying the login info
			makeGetRequest('/api/account?space=' + accountSpace + '&username='+ lUsername + '&password=' + lPassword, onSuccess, onFailure);
		});
	};
	
	
	var attachCreateAccountHandler = function(e){
		
		$(".account-creation-form").on('click', ".signup-button", function(e) {
			
			e.preventDefault ();	// Tell the browser to skip its default click action
			
			if ($("input[name=account-type]:checked").val() == "student"){	//if the radio button's value is 'student', create a student account
				createStudentAccount();
			}
			else if ($("input[name=account-type]:checked").val() == "instructor"){ //if the radio button's value is 'instructor', create an instructor account
				createInstructorAccount();
			}
		});
	};
	
	var attachPrefixDropdownTextHandler = function(){
		$("#course-prefix-dropdown").on('click', 'li a', function(){
			$("#selected-prefix:first-child").text($(this).text());
			$("#selected-prefix:first-child").val($(this).text());
		});
	};
	
	var attachLogoutListener = function(){
		$("#logout-button").click(function(){
			window.localStorage.clear();
			window.location.href = "index.html";	//if we successfully created an account, go back to the login page
		});
	};
	
	//	Waits until the page is loaded before running these functions.
	$(document).ready(function(){
		resetRadios();
		attachStudentFieldDisplayListener();
		attachConfirmPasswordListener();
		attachCreateAccountHandler();
		attachLoginHandler();
		attachPrefixDropdownTextHandler();
		attachLogoutListener();
		
		homeLoadUserData();
		accountLoadUserData();
		
		
	});
	
})();